### PR TITLE
txn: distinguish write conflict reasons

### LINKED ISSUE #37734: Distinguish different types of write conflicts
## Enhancement
Write conflicts errors are returnred for different purposes, we need to distingush them to that the error gives a clear message.

### PR DESCRIPTION
**Pull request by ekexium**
### What problem does this PR solve?

Issue Number: close #37734 

Problem Summary:

Write conflicts errors are returnred for different purposes, we need to distingush them to that the error gives a clear message.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Distinguish different types of write conflict
```

### CODE REVIEW

**Comment by ekexium** [comment]
Let's wait https://github.com/tikv/client-go/pull/584 and merge it together.

File: tests/realtikvtest/pessimistictest/pessimistic_test.go [hunk 1]
```diff
@@ -3237,14 +3237,15 @@ func TestLazyUniquenessCheck(t *testing.T) {
 	tk.MustQuery("select * from t2").Check(testkit.Rows("1 12"))
 	tk.MustExec("admin check table t")
 
-	// case: conflict check failure
+	// case: conflict check failure, it doesn't commit in the optimistic way though there is no lock acquired in the txn
```
Comment (TonsnakeLin) [inline tests/realtikvtest/pessimistictest/pessimistic_test.go:3240]: It's better to add more test cases for other conflict reasons.
The reset LGTM

**Review by cfzjywxk — APPROVED** [review]
LGTM

**Review by sticnarf — APPROVED** [review]
LGTM

**Pull request merged**